### DISCUSSION TITLE
Alternative Python linkage

### DISCUSSION THREAD

**Discussion by airallergy** [NREL/EnergyPlus#9188] [Q&A] [answered]
Hi,

To use Python Plugin, I am wondering if I can link an alternative python executable instead of the one shipped with EnergyPlus (without compiling from source codes)?

For some reason, I have to use EP 9.4 (which contains Python 3.7) and Python 3.8, which makes it tricky to include third-party modules like numpy as per my current understanding. I saw that you are thinking of making [python-free distributions](https://github.com/NREL/EnergyPlus/issues/9083), so I guess there might be a way that I failed to find?

**Comment by Myoldmopar** [comment, marked as answer]
I would highly suggest not actually using the Python Plugin system in this case, but rather call EnergyPlus through the Python API.  The approach is actually quite similar, the fundamental being that instead of relying on a packaged EnergyPlus Python version, you simply call the EnergyPlus API from your own Python install.  This way you can use any Python version you want, and any extra libraries you want.  

We do have a few examples of this, such as here: https://github.com/NREL/EnergyPlus/blob/develop/tst/EnergyPlus/api/TestRuntime.py#L77.  The steps are basically:

- Change your Python Plugin class method into a callback function in your Python file
- Create an EnergyPlus state instance
- Register your callback function to be called back at the same calling point that you would have used for Python Plugins
- Call the EnergyPlus "function" to execute EnergyPlus, and EnergyPlus will call your callback function while the simulation is running.

The contents of your callback function can do all the things you'd expect to do, including making data exchange API calls with the running simulation to read sensors and write actuators.

**Reply by airallergy** [reply]
@Myoldmopar, thanks for your thorough instructions, I got to play with the api method a bit and it works like a charm. I have two further questions about the api:

1. I am currently using `sys.path.append("/Applications/EnergyPlus-9-4-0")` at the beginning of the script to let python find `pyenergyplus`, is this a recommended way to do so?
2. When I tried to run with `python *.py -r -w *.epw *.idf`, EP tried to find`ReadVarsESO` in my python env, which gave an apparent error `ERROR: Could not find ReadVarsESO executable: /Users/Cheng/mambaforge/envs/bps/bin/PostProcess/ReadVarsESO.`. Any thoughts?

**Reply by Myoldmopar** [reply]
Glad it works well!

1. Yes that is totally fine.  You could also pass in the E+ directory as an argument if that worked for your workflow, that way it wouldn't be hardcoded in your script.  You could also just set PYTHONPATH as an environment variable, and Python will find it based on that.  Whatever works to tell Python where to find the E+ library.
2. Yeah, that error is fixed now, but the fix isn't until 9.6.  For now, just don't try to call ReadVars using the -r flag.  Since you are already calling E+ from your own Python script, just execute ReadVars manually as a Python subprocess.

**Reply by airallergy** [reply]
Great to know it is fixed in 9.6! I was executing readvars manually instead, just wondering where the issue stems from. Many thanks!
